src/routes/settings/components/Settings.js

```diff
@@ -0,0 +1,235 @@
+const React = require('react');
+const T = require('prop-types');
+const NavLink = require('react-router-dom').NavLink;
+const StrangeForms = require('strange-forms');
+const IsEmail = require('utils/is-email');
+const { Button, TextField, FormHelperText, Divider, Typography } = require('@material-ui/core');
+const { FormWrapper, TextWrapper, ButtonWrapper } = require('styles/global-components.js');
+
+
+module.exports = class Settings extends StrangeForms(React.Component) {
+
+    static propTypes = {
+        onSubmit: T.func.isRequired,
+        errorMessage: T.string,
+        userDetails: T.shape({
+            firstName: T.string.isRequired,
+            lastName: T.string.isRequired,
+            email: T.string.isRequired
+        })
+    };
+
+    constructor(props) {
+
+        super(props);
+
+        this.state = {
+            firstName: props.userDetails.firstName || '',
+            lastName: props.userDetails.lastName || '',
+            email: props.userDetails.email || '',
+            currentPassword: '',
+            password: '',
+            confirmPassword: '',
+            isBlurred: {
+                email: false,
+                confirmPassword: false
+            }
+        };
+
+        this.strangeForm({
+            fields: ['firstName', 'lastName', 'email', 'currentPassword', 'password', 'confirmPassword'],
+            get: (someProps, field) => this.state[field],
+            act: (field, value) => this.setState({ [field]: value }),
+            getFormValue: {
+                '*': this.getFormValue.bind(this)
+            }
+        });
+
+        this.fieldBlurred = this._fieldBlurred.bind(this);
+        this.showEmailError = this._showEmailError.bind(this);
+        this.showPasswordError = this._showPasswordError.bind(this);
+        this.disableButton = this._disableButton.bind(this);
+        this.submit = this._submit.bind(this);
+    }
+
+    getCheckedValue(e) {
+
+        return e.target.checked;
+    }
+
+    getFormValue(e) {
+
+        return e.target.value || '';
+    }
+
+    _fieldBlurred(ev) {
+
+        const isBlurred = { ...this.state.isBlurred };
+        const field = ev.target.id;
+
+        isBlurred[field] = true;
+
+        this.setState({ isBlurred });
+    }
+
+    _showEmailError() {
+
+        return (this.state.isBlurred.email) && !IsEmail(this.state.email);
+    }
+
+    _showPasswordError() {
+
+        return this.state.isBlurred.confirmPassword && (this.state.password !== this.state.confirmPassword);
+    }
+
+    _disableButton() {
+
+        const { firstName, lastName, email, currentPassword, password, confirmPassword } = this.state;
+        // const fieldHasValue = (firstName && lastName && email && password) !== '';
+        const fieldHasChanged =
+            (firstName !== this.props.userDetails.firstName)
+            || (lastName !== this.props.userDetails.lastName)
```

Comment:
It's a total nit (or not even, since it's purely arbitrary) but you'll typically see BRS code have multi-line conditions written like this,
```js
        const fieldHasChanged = (firstName !== this.props.userDetails.firstName) ||
            (lastName !== this.props.userDetails.lastName) ||
            (email !== this.props.userDetails.email);
```

Reply:
Oh, weird! I can change it to match the BRS convention if you prefer, but I'm partial to seeing the operation at the start of the line. To me, the difference is like

>123 x
4

compared to

>123
x 4

Reply:
Yes, please!  We go by the hapi style guide, which has a rule about this: https://hapijs.com/styleguide#operators

Maybe it goes without saying, but some things that look weird to one person wont look weird to another.  Both of those examples above feel equally ergonomic to me.  A few people here felt really weird moving to [Stroustrup curlies](https://en.wikipedia.org/wiki/Indentation_style#Variant:_Stroustrup), for example, but over time it felt natural.

Reply:
I'll make it so!

And totally—I got over using four spaces instead of two, so this is no big deal :)